server/permalinks.go

```diff
@@ -0,0 +1,107 @@
+	word       string            // the permalink
+	captureMap map[string]string // named regex capture group of the link
+}
+
+// getReplacements returns the permalink replacements that needs to be performed
+// on a message.
+func (p *Plugin) getReplacements(msg string) []replacement {
+	// find the permalinks from the msg using a regex
+	matches := p.githubPermalinkRegex.FindAllStringSubmatch(msg, -1)
+	indices := p.githubPermalinkRegex.FindAllStringIndex(msg, -1)
+	replacements := make([]replacement, 0, maxPermalinkReplacements)
+	for i, m := range matches {
+		// have a limit on the no. of replacements to do
+		if i > maxPermalinkReplacements {
+			break
+		}
+		word := m[0]
+		index := indices[i][0]
+		r := replacement{
+			index: index,
+			word:  word,
+		}
+		// ignore if the word is inside a link
+		if isInsideLink(msg, index) {
+			continue
+		}
+		// populate the captureMap with the extracted groups of the regex
+		r.captureMap = make(map[string]string)
+		for j, name := range p.githubPermalinkRegex.SubexpNames() {
+			if j == 0 {
+				continue
+			}
+			r.captureMap[name] = m[j]
+		}
+		replacements = append(replacements, r)
+	}
+	return replacements
+}
+
+// makeReplacements perform the given replacements on the msg and returns
+// the new msg.
+func (p *Plugin) makeReplacements(msg string, replacements []replacement) string {
+	for _, r := range replacements {
+		// quick bailout if the commit hash is not proper
+		if _, err := hex.DecodeString(r.captureMap["commit"]); err != nil {
+			p.API.LogError("bad git commit hash in permalink", "error", err.Error(), "hash", r.captureMap["commit"])
+			continue
+		}
+
+		// get the file contents
+		opts := github.RepositoryContentGetOptions{
+			Ref: r.captureMap["commit"],
+		}
+		// TODO: make all of these requests concurrently.
+		reqctx, cancel := context.WithTimeout(context.Background(), permalinkReqTimeout)
+		fileContent, _, _, err := p.githubClient.Repositories.GetContents(reqctx,
+			r.captureMap["user"], r.captureMap["repo"], r.captureMap["path"], &opts)
+		if err != nil {
+			p.API.LogError("error while fetching file contents", "error", err.Error(), "path", r.captureMap["path"])
+			cancel()
+			continue
+		}
+		cancel()
+		// this is not a file, ignore.
+		if fileContent == nil {
+			p.API.LogWarn("permalink is not a file", "file", r.captureMap["path"])
+			continue
+		}
+		decoded, err := fileContent.GetContent()
+		if err != nil {
+			p.API.LogError("error while decoding file contents", "error", err.Error(), "path", r.captureMap["path"])
+			continue
+		}
+
+		// get the required lines
+		start, end := getLineNumbers(r.captureMap["line"])
+		// bad anchor tag, ignore.
+		if start == -1 || end == -1 {
+			continue
+		}
+		lines := filterLines(decoded, start, end, maxPreviewLines)
+		final := getCodeMarkdown(r.captureMap, r.word, lines, start, end)
+
+		// replace word in msg starting from r.index only once
+		msg = msg[:r.index] + strings.Replace(msg[r.index:], r.word, final, 1)
```

Comment:
Hmmm, after updating the text here, wouldn't `r.index` become invalid for the next iteration as things have moved? Unless maybe I'm missing something... 🤔 

One way to fix this would be to always do the replacement starting from the right-most index and work your way in towards the start of the string. That way the indexes are valid across all replacements.

Reply:
> Hmmm, after updating the text here, wouldn't r.index become invalid for the next iteration as things have moved?

Yeah :man_facepalming:  .. I was also stuck trying to understand how is it working. The reason is because it is trying to replace the string just once. So even if the starting index is incorrect. It just needs the string to be present in the entire string. So accidentally, this worked :sweat_smile: 

Fixed it to do it in reverse.